src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -0,0 +1,250 @@
+			}
+
+			_batchBlockReceiveAsyncCts.Cancel();
+
+			try
+			{
+				await _worker;
+			}
+			catch (TaskCanceledException)
+			{
+				_logger.LogDebug("worker task cancelled");
+			}
+
+			_eventQueue.Complete();
+		}
+
+		private async Task DoWork()
+		{
+			_batchBlockReceiveAsyncCts = new CancellationTokenSource();
+			while (_isProcessingLoopRunning)
+			{
+				var queueItems = await _eventQueue.ReceiveAsync(_batchBlockReceiveAsyncCts.Token);
+				await ProcessQueueItems(queueItems);
+			}
+		}
+
+		private async Task ProcessQueueItems(object[] queueItems)
+		{
+			try
+			{
+				var metadata = new Metadata { Service = _service };
+				var metadataJson = JsonConvert.SerializeObject(metadata, _settings);
+				var json = new StringBuilder();
+				json.Append("{\"metadata\": " + metadataJson + "}" + "\n");
+
+				foreach (var item in queueItems)
+				{
+					var serialized = JsonConvert.SerializeObject(item, _settings);
+					switch (item)
+					{
+						case Transaction _:
+							json.AppendLine("{\"transaction\": " + serialized + "}");
+							break;
+						case Span _:
+							json.AppendLine("{\"span\": " + serialized + "}");
+							break;
+						case Error _:
+							json.AppendLine("{\"error\": " + serialized + "}");
+							break;
+					}
+				}
+
+				var content = new StringContent(json.ToString(), Encoding.UTF8, "application/x-ndjson");
+
+				var result = await _httpClient.PostAsync(Consts.IntakeV2Events, content);
+
+				if (result != null && !result.IsSuccessStatusCode)
+				{
+					var str = await result.Content.ReadAsStringAsync();
+					_logger.LogError($"Failed sending event. {str}");
+				}
+				if (result != null && result.IsSuccessStatusCode)
+				{
+					var sb = new StringBuilder();
+					sb.AppendLine("Sent items to server:");
+					foreach (var item in queueItems) sb.AppendLine(item.ToString());
+					_logger.LogDebug(sb.ToString());
+				}
+			}
+			catch (Exception e)
+			{
+				var sb = new StringBuilder();
+				sb.AppendLine("Failed sending events. ");
+				sb.Append("Exception: ");
+				sb.Append(e.GetType().FullName);
+				sb.Append(", Message: ");
+				sb.AppendLine(e.Message);
+				sb.AppendLine("Following events were not transferred successfully to the server:");
+				foreach (var item in queueItems) sb.AppendLine(item.ToString());
+
+				_logger.LogWarning(sb.ToString());
+			}
+		}
+	}
+
+	internal class Metadata
+	{
+		// ReSharper disable once UnusedAutoPropertyAccessor.Global - used by Json.Net
+		public Service Service { get; set; }
+	}
+
+	//Credit: https://stackoverflow.com/a/30726903/1783306
+	internal sealed class SingleThreadTaskScheduler : TaskScheduler
+	{
+		[ThreadStatic]
+		private static bool _isExecuting;
+
+		private readonly CancellationToken _cancellationToken;
+
+		private readonly BlockingCollection<Task> _taskQueue;
+
+		public SingleThreadTaskScheduler(CancellationToken cancellationToken)
+		{
+			_cancellationToken = cancellationToken;
+			_taskQueue = new BlockingCollection<Task>();
```

Comment:
Maybe it's worth giving the thread more descriptive name like "Elastic APM PayloadSender"?

Reply:
Good idea! Changed it.